server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -213,9 +214,16 @@ public RepositoryData addSnapshot(final SnapshotId snapshotId,
         newSnapshotStates.put(snapshotId.getUUID(), snapshotState);
         Map<String, Version> newSnapshotVersions = new HashMap<>(snapshotVersions);
         newSnapshotVersions.put(snapshotId.getUUID(), version);
-        Map<IndexId, Set<SnapshotId>> allIndexSnapshots = new HashMap<>(indexSnapshots);
+        Map<IndexId, List<SnapshotId>> allIndexSnapshots = new HashMap<>(indexSnapshots);
         for (final IndexId indexId : shardGenerations.indices()) {
-            allIndexSnapshots.computeIfAbsent(indexId, k -> new LinkedHashSet<>()).add(snapshotId);
+            final List<SnapshotId> snapshotIds = allIndexSnapshots.get(indexId);
+            if (snapshotIds == null) {
+                allIndexSnapshots.put(indexId, List.of(snapshotId));
+            } else {
+                final List<SnapshotId> copy = new ArrayList<>(snapshotIds);
+                copy.add(snapshotId);
+                allIndexSnapshots.put(indexId, List.copyOf(copy));
```

Comment:
why create a copy of the copy?

Reply:
These `RepositoryData` instances live for quite a while, so I figured the cost of doing another copy is worth the lower storage overhead + shorter path to the GC root compared to wrapping with `Collections.unmodifiableList`? I could technically make this more efficient by copying to a `SnapshotId[]` and then just wrapping that array but I figured this wasn't that much slower and nicer to read.

Reply:
I looked into how `List.copyOf` is implemented, and lo and behold, it copies the elements twice (first calls `Collection.toArray()`, and then creates another copy of that temporary array in `List.of` (using manual for loop, FFS).
This means that the list is copied three times here, plus the resize of the `ArrayList` when calling `copy.add(snapshotId);`, leading to another full copy ....
High-level languages ftw.

Reply:
:) you win => I pushed https://github.com/elastic/elasticsearch/pull/55293/commits/8319e2116d138668e38c3a72b709a8889868088a , probably not worth the hassle to go further than this then.